src/Controller/User/Profile/User2FAController.php

```diff
@@ -174,27 +176,38 @@ private function handleForm(
     ): FormInterface|Response {
         $form->handleRequest($request);
 
-        if ($form->isSubmitted()
-                && $form->has('totpCode')
+        if (!$form->isSubmitted()) {
+            return $form;
+        }
+
+        if ($form->has('totpCode')
                 && !$this->setupHasValidCode($dto->totpSecret, $form->get('totpCode')->getData())) {
             $form->get('totpCode')->addError(new FormError($this->translator->trans('2fa.code_invalid')));
-        }
 
-        if ($form->isSubmitted() && $form->isValid()) {
-            $this->manager->edit($this->getUser(), $dto);
+            return $form;
+        }
 
-            if ($dto->totpSecret) {
-                $this->security->logout(false);
+        if (!$form->isValid()) {
+            $this->logger->warning('2fa error occurred user "{username}" submitting the form "{errors}"', [
+                'username' => $dto->username,
+                'errors' => $form->getErrors(),
+            ]);
+            $form->get('totpCode')->addError(new FormError($this->translator->trans('2fa.setup_error')));
 
-                $this->addFlash('success', 'flash_account_settings_changed');
+            return $form;
+        }
 
-                return $this->redirectToRoute('app_login');
-            }
+        $this->manager->edit($this->getUser(), $dto);
 
+        if (!$dto->totpSecret) {
```

Comment:
I couldn't quite figure out in which cases this was true, maybe in DB errors or something if it can't edit the user? I left it as is with the redirect for now